File: app/src/main/java/cz/vutbr/fit/brnogo/tools/DiffUtilCallback.java
package cz.vutbr.fit.brnogo.tools;

import android.support.v7.util.DiffUtil;

import java.util.List;

import cz.vutbr.fit.brnogo.data.model.recyclerview.AdapterItem;

public class DiffUtilCallback<T extends AdapterItem> extends DiffUtil.Callback {

	private List<T> oldItems;
	private List<T> newItems;

	public DiffUtilCallback(List<T> oldItems, List<T> newItems) {
		this.oldItems = oldItems;
		this.newItems = newItems;
	}

	@Override
	public int getOldListSize() {
		return oldItems.size();
	}

	@Override
	public int getNewListSize() {
		return newItems.size();
	}

	@Override
	public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
		return oldItems.get(oldItemPosition).getId().equals(newItems.get(newItemPosition).getId());
	}

	@Override
	public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
		return oldItems.get(oldItemPosition).equals(newItems.get(newItemPosition));
	}
}

File: app/src/main/java/cz/vutbr/fit/brnogo/tools/DiffUtilCallback.kt
package cz.vutbr.fit.brnogo.tools

import android.support.v7.util.DiffUtil

import cz.vutbr.fit.brnogo.data.model.recyclerview.AdapterItem

/**
 * Utility class that can calculate the difference between two lists and output.
 * It is used to calculate updates for a RecyclerView Adapter.
 */

class DiffUtilCallback<T : AdapterItem>(private val oldItems: List<T>, private val newItems: List<T>) : DiffUtil.Callback() {

	override fun getOldListSize(): Int {
		return oldItems.size
	}

	override fun getNewListSize(): Int {
		return newItems.size
	}

	override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
		return oldItems[oldItemPosition].itemId == newItems[newItemPosition].itemId
	}

	override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
		return oldItems[oldItemPosition] == newItems[newItemPosition]
	}
}
